### PR TITLE
Fix compiler warning

### PR DESCRIPTION
**Pull request by jiaxiyan**
The compiler gives warning: 'cid_base' may be used uninitialized in this function.

### CODE REVIEW

**Comment by dalcinl** [comment]
mpi4py failure... this one is unrelated to this PR, and already reported in #12367
```
testAlltoallInter (test_util_pkl5.TestMPIWorld.testAlltoallInter) ... python: pml_ob1_comm.c:51: mca_pml_ob1_comm_proc_destruct: Assertion `NULL == proc->frags_cant_match' failed.
```

More disturbing, I re-ran the testsuite with re-enabled ULFM tests again, and now I get a [failure](https://github.com/mpi4py/mpi4py-testing/actions/runs/8097420010/job/22201106919). Not sure whether this is a consequence of applying @abouteiller suggestion or something else. I'll have to test this locally before proceeding, therefore I'm marking this PR as draft.

**Comment by dalcinl** [comment]
I posted and already deleted a comment in this PR. That comment belonged elsewhere. Sorry for the noise.

File: ompi/communicator/comm_cid.c [hunk 1]
```diff
@@ -315,7 +315,7 @@ static int ompi_comm_ext_cid_new_block (ompi_communicator_t *newcomm, ompi_commu
     opal_process_name_t *name_array = NULL;
     char *tag = NULL;
     size_t proc_count;
-    size_t cid_base;
+    size_t cid_base = 0;
```
Comment (bosilca) [inline ompi/communicator/comm_cid.c:318]: This fixes the warning but will break the code. Fact: we cannot have two CID blocks with the same base, or we will create conflicting set of CIDs (and each communicator is expected to have a unique CID). Thus, if the cid_base is not correctly extracted from the PMIX group we shall not continue, resulting in creating a second block with the same default (aka 0 with this patch) CID.

Comment (dalcinl) [reply ompi/communicator/comm_cid.c:318]: @bosilca I belive you have misread the proposed fix. The act of just initializing a previously uninitialized variable cannot break the code by itself unless the initial value wrong. If initializing `cid_base = 0` is not the right initial value, then which value should the OP use?

Comment (dalcinl) [reply ompi/communicator/comm_cid.c:318]: Maybe initialize  `cid_base = -1;`, and right after that value is used for read add `assert (cid_base >= 0);` ?

Comment (bosilca) [reply ompi/communicator/comm_cid.c:318]: I do believe it can, deterministically lead to faults. While the current code might work, just not deterministically. I explained the reasons above, if we create a new block with a random base (which will happen most of the time is we do explicitly initialize the variable), the code might seems to be working just fine. 

I'm not advocating on keeping the current code, I'm just pointing out that this fix might have a worst outcome than the original.

Comment (wenduwan) [reply ompi/communicator/comm_cid.c:318]: To my understanding, an uninitialized variable does not really get a true random value. So as George said, are we lucky that the code works as-is?

@jiaxiyan Did you test this PR? Any failures?

Comment (wenduwan) [reply ompi/communicator/comm_cid.c:318]: Reading the code, `cid_base` should be assigned on L383
```
    for (size_t i=0; i<nresults; i++) {
        if (PMIX_CHECK_KEY(&results[i], PMIX_GROUP_CONTEXT_ID)) {
            PMIX_VALUE_GET_NUMBER(rc, &results[i].value, cid_base, size_t);
            if(PMIX_SUCCESS != rc) {
                ret = opal_pmix_convert_status(rc);
                goto fn_exit;
            }
            break;
        }
    }
```

I wonder if it's possible that we don't hit this path and end up with the default value - would that be erroneous anyways?

Comment (jsquyres) [reply ompi/communicator/comm_cid.c:318]: @bosilca If `PMIX_VALUE_GET_NUMBER()` fails to set `cid_base`, won't we go to `fn_exit` and therefore ignore the `cid_base` value?  I don't think it's possible for `nresults` to be less than 1, right?

If that's the case, then @dalcinl's suggestion of initially assigning the value to `-1` is probably good -- it makes it obvious that the value is invalid, but we'll only end up with `cid_base==-1` if `PMIX_VALUE_GET_NUMBER()` fails and we end up going to `fn_exit`.

Comment (jiaxiyan) [reply ompi/communicator/comm_cid.c]: Updated initialization to -1 as suggested

Comment (wenduwan) [reply ompi/communicator/comm_cid.c]: I'm not sure about Open MPI but some people prefer `SIZE_MAX` over `-1` for `size_t`.

Also this solution is not complete - you also need to check `cid_base` value and error out if it stays unassigned.

Comment (dalcinl) [reply ompi/communicator/comm_cid.c]: @jiaxiyan Could you please add a line with

```c
assert (cid_base >= 0);
```
right BEFORE the line with
```c
 ompi_comm_extended_cid_block_initialize (new_block, cid_base, 0, 0);
```
That way, if `cid_base` is used but never set to anything meaningful, at least we know where to point our fingers.

Comment (wenduwan) [reply ompi/communicator/comm_cid.c]: @dalcinl Note that `cid_base` is unsigned - I'm not sure `assert (cid_base >= 0);` will do anything.

The other option is to use a pointer `size_t *cid_base_ptr = NULL;` and only initialize it if `cid_base` is set correctly. Then we can do a NULL check on `cid_base_ptr`.

Comment (dalcinl) [reply ompi/communicator/comm_cid.c]: Sorry, it is `size_t`!, then yes, it would be better to 

```c
size_t cid_base  = SIZE_MAX;
```
and then
```c
assert(SIZE_MAX != cid_base)
```
Would that work?

Comment (jsquyres) [reply ompi/communicator/comm_cid.c]: If there's no meaningful sentinel value (e.g., because it's unsigned), you could just add another variable, such as `bool cid_base_set = false`, and only set it to `true` when the `PMIX_VALUE_GET_NUMBER()` returns successfully.  Then you can check `cid_base_set` before calling `ompi_comm_extend_cid_block_initialize()`.

It's not the most elegant thing in the world, but it is effective, easy to read/grok, and 100% safe.

Comment (dalcinl) [reply ompi/communicator/comm_cid.c]: > Then you can check `cid_base_set`

If not, would it make sense to do cleanup and error with let say `ERR_INTERN` or `OMPI_ERR_OUT_OF_RESOURCE` ? Better than a naive `assert()`, right?

Comment (jsquyres) [reply ompi/communicator/comm_cid.c]: Yes, we should not do `assert()`.  Realistically, if `PMIX_VALUE_GET_NUMBER()` fails, the current code *already* handles that case:

```c
    for (size_t i=0; i<nresults; i++) {
        if (PMIX_CHECK_KEY(&results[i], PMIX_GROUP_CONTEXT_ID)) {
            PMIX_VALUE_GET_NUMBER(rc, &results[i].value, cid_base, size_t);
            if(PMIX_SUCCESS != rc) {
                ret = opal_pmix_convert_status(rc);
                goto fn_exit;
            }
            break;
        }
    }
```

I'm guessing the compiler is just not smart enough / does not have enough context to know that we should *always* execute `PMIX_VALUE_GET_NUMBER()` here.

That being said, adding a belt-and-suspenders check for ensure that `cid_base` is *actually* set -- although likely redundant -- isn't harmful, and adds a bit of defensiveness in the code, which is not a bad thing.

Comment (rhc54) [reply ompi/communicator/comm_cid.c]: You really don't want to use an `assert` here as that only applies during a debug build, and this problem is just as likely to hit in non-debug scenarios. The only real way you would encounter it is in an environment that does not support `PMIx_Group` operations - which means a direct launch scenario like Slurm. In those cases, you'll see an error returned from the group op, and you'll know that the op failed and cid_base wasn't set by just looking at the return code.

In other words, all this is pretty much wasted breath, guys. You check the return from `PMIx_Group_construct`, and if it is an error, then you go to the error return. If not, then you check the error code from `PMIX_VALUE_GET_NUMBER` - and if that is success, then the cid_base is good.

You still need to init the variable to shut up the compiler, but all the angst is unnecessary.

Comment (rhc54) [reply ompi/communicator/comm_cid.c]: `PMIX_VALUE_GET_NUMBER` is a macro:
```c
#define PMIX_VALUE_GET_NUMBER(s, m, n, t)               \
    do {                                                \
        (s) = PMIX_SUCCESS;                             \
        if (PMIX_SIZE == (m)->type) {                   \
            (n) = (t)((m)->data.size);                  \
        } else if (PMIX_INT == (m)->type) {             \
            (n) = (t)((m)->data.integer);               \
        } else if (PMIX_INT8 == (m)->type) {            \
            (n) = (t)((m)->data.int8);                  \
        } else if (PMIX_INT16 == (m)->type) {           \
            (n) = (t)((m)->data.int16);                 \
        } else if (PMIX_INT32 == (m)->type) {           \
            (n) = (t)((m)->data.int32);                 \
        } else if (PMIX_INT64 == (m)->type) {           \
            (n) = (t)((m)->data.int64);                 \
        } else if (PMIX_UINT == (m)->type) {            \
            (n) = (t)((m)->data.uint);                  \
        } else if (PMIX_UINT8 == (m)->type) {           \
            (n) = (t)((m)->data.uint8);                 \
        } else if (PMIX_UINT16 == (m)->type) {          \
            (n) = (t)((m)->data.uint16);                \
        } else if (PMIX_UINT32 == (m)->type) {          \
            (n) = (t)((m)->data.uint32);                \
        } else if (PMIX_UINT64 == (m)->type) {          \
            (n) = (t)((m)->data.uint64);                \
        } else if (PMIX_FLOAT == (m)->type) {           \
            (n) = (t)((m)->data.fval);                  \
        } else if (PMIX_DOUBLE == (m)->type) {          \
            (n) = (t)((m)->data.dval);                  \
        } else if (PMIX_PID == (m)->type) {             \
            (n) = (t)((m)->data.pid);                   \
        } else if (PMIX_PROC_RANK == (m)->type) {       \
            (n) = (t)((m)->data.rank);                  \
        } else if (PMIX_STATUS == (m)->type) {          \
            (n) = (t)((m)->data.status);                \
        } else {                                        \
            (s) = PMIX_ERR_BAD_PARAM;                   \
        }                                               \
    } while(0)
```
As you can see, it returns an error if the number cannot be provided. It does _not_ set the number to a value in that case because we have no idea what value would constitute an error. So only the error code is relevant.

Compiler is just flagging that the value isn't set in an error case. Question is: how is `cid_base` being used in that situation...and why?

Comment (bosilca) [reply ompi/communicator/comm_cid.c]: @rhc all error case are correctly handled. And no the compiler does not complain about the value not being set in the error cases, because error cases lead to `fn_exit` with no use of `cid_base`.

The compiler complains about the case where `nresults` is zero or no returned info matches PMIX_GROUP_CONTEXT_ID, so cid_base remains uninitialized when the code reaches `ompi_comm_extended_cid_block_initialize`.

In that case setting cid_base to a predefined value will lead to more or less badness: from the worst, aka. colliding blocks of cid, to the best, aka. a steep performance degradation increasing as we create more communicators due to the need to find a cid available everywhere (cid that starts with a default predefined value, so we beat the same cid range all the time).

The current code, with the non initialized cid_base, has a better chance of success, because at least one of the processes might bring a different value. The solution to go for a default value could work for as long as we inform the user that we are defaulting to a predefined value and that might have significant performance implications on communicator creation.

Comment (rhc54) [reply ompi/communicator/comm_cid.c]: Hmmm...I see. Of course, those conditions (nresults=0 or no returned context ID, but group construct is successful) cannot happen, but that isn't something the compiler can know.

It's hard to find a value you can set to indicate invalid. Per request of @hppritcha, PRRTE starts at `UINT32_MAX` and counts downward. IIRC, somewhere OMPI starts at 0 and counts upward?

Sounds like @jsquyres suggestion is probably the best - use a boolean to indicate that a value was found. If not found, then perhaps the best solution is to have each proc generate its own random number? Would solve the "same range" issue, though I have no idea of what the overall impact might be.

File: ompi/communicator/comm_cid.c [hunk 2]
```diff
@@ -395,6 +397,11 @@ static int ompi_comm_ext_cid_new_block (ompi_communicator_t *newcomm, ompi_commu
         goto fn_exit;
     }
 
+    if (!cid_base_set) {
+        opal_output_verbose (0, 0, "cid_base value is not set\n");
```
Comment (jsquyres) [inline ompi/communicator/comm_cid.c:401]: There should not be a trailing `\n` in `opal_output_verbose()` -- it will line terminate for you.

That being said, I think we should probably `opal_show_help()` here, not just a single line output.  One of the big reasons for this is that `opal_show_help()` output is de-duplicated back in `mpirun` output (e.g., if 10,000 MPI processes emit this message, it will only be displayed once with an indication that 9,999 other processes also emitted this message).  There should probably be an appropriate help message that "This should probably never happen ... etc."

**Comment by jsquyres** [comment]
@jiaxiyan Thanks for updating!  I pushed a minor tweak to your show_help message, and I think we're good to go.

**Review by jsquyres — APPROVED** [review]

**Comment by wenduwan** [comment]
mpi4py failure. It's concerning that `ompi_comm_nextcid` is in the backtrace.
```
testErrcodes (test_spawn.TestSpawnMultipleWorld.testErrcodes) ... testErrcodes (test_spawn.TestSpawnMultipleWorld.testErrcodes) ... [fv-az702-107:145232] *** Process received signal ***
[fv-az702-107:145232] Signal: Segmentation fault (11)
[fv-az702-107:145232] Signal code: Address not mapped (1)
[fv-az702-107:145232] Failing at address: 0x180
[fv-az702-107:145232] [ 0] /lib/x86_64-linux-gnu/libc.so.6(+0x42520)[0x7fe163242520]
[fv-az702-107:145232] [ 1] /usr/local/lib/libmpi.so.0(+0x399eaf)[0x7fe162799eaf]
[fv-az702-107:145232] [ 2] /usr/local/lib/libmpi.so.0(mca_pml_ob1_recv_frag_callback_match+0x169)[0x7fe16279c21e]
[fv-az702-107:145232] [ 3] /usr/local/lib/libopen-pal.so.0(+0xcc700)[0x7fe162382700]
[fv-az702-107:145232] [ 4] /usr/local/lib/libevent_core-2.1.so.7(+0x1e3a8)[0x7fe1634a43a8]
[fv-az702-107:145232] [ 5] /usr/local/lib/libevent_core-2.1.so.7(event_base_loop+0x577)[0x7fe1634a4b07]
[fv-az702-107:145232] [ 6] /usr/local/lib/libopen-pal.so.0(+0x267fc)[0x7fe1622dc7fc]
[fv-az702-107:145232] [ 7] /usr/local/lib/libopen-pal.so.0(opal_progress+0xab)[0x7fe1622dc8b9]
[fv-az702-107:145232] [ 8] /usr/local/lib/libopen-pal.so.0(ompi_sync_wait_mt+0x1d1)[0x7fe16232b284]
[fv-az702-107:145232] [ 9] /usr/local/lib/libmpi.so.0(+0x70807)[0x7fe162470807]
[fv-az702-107:145232] [10] /usr/local/lib/libmpi.so.0(ompi_comm_nextcid+0x8d)[0x7fe162472710]
[fv-az702-107:145232] [11] /usr/local/lib/libmpi.so.0(ompi_dpm_connect_accept+0x3f25)[0x7fe162485085]
[fv-az702-107:145232] [12] /usr/local/lib/libmpi.so.0(ompi_dpm_dyn_init+0xd9)[0x7fe16248f53b]
[fv-az702-107:145232] [13] /usr/local/lib/libmpi.so.0(ompi_mpi_init+0x83b)[0x7fe1624b26f1]
[fv-az702-107:145232] [14] /usr/local/lib/libmpi.so.0(PMPI_Init_thread+0xe0)[0x7fe16251d0ea]
[fv-az702-107:145232] [15] /home/runner/work/ompi/ompi/build/lib.linux-x86_64-cpython-312/mpi4py/MPI.cpython-312-x86_64-linux-gnu.so(+0x3dbaf)[0x7fe162a3dbaf]
[fv-az702-107:145232] [16] /home/runner/work/ompi/ompi/build/lib.linux-x86_64-cpython-312/mpi4py/MPI.cpython-312-x86_64-linux-gnu.so(+0x26d396)[0x7fe162c6d396]
[fv-az702-107:145232] [17] /opt/hostedtoolcache/Python/3.12.2/x64/lib/libpython3.12.so.1.0(PyModule_ExecDef+0x69)[0x7fe1638df8a9]
[fv-az702-107:145232] [18] /opt/hostedtoolcache/Python/3.12.2/x64/lib/libpython3.12.so.1.0(+0x31d660)[0x7fe16391d660]
[fv-az702-107:145232] [19] /opt/hostedtoolcache/Python/3.12.2/x64/lib/libpython3.12.so.1.0(+0x241f77)[0x7fe163841f77]
[fv-az702-107:145232] [20] /opt/hostedtoolcache/Python/3.12.2/x64/lib/libpython3.12.so.1.0(_PyEval_EvalFrameDefault+0x4f0e)[0x7fe163880e4e]
[fv-az702-107:145232] [21] /opt/hostedtoolcache/Python/3.12.2/x64/lib/libpython3.12.so.1.0(+0x219d35)[0x7fe163819d35]
[fv-az702-107:145232] [22] /opt/hostedtoolcache/Python/3.12.2/x64/lib/libpython3.12.so.1.0(PyObject_CallMethodObjArgs+0x103)[0x7fe16381ae43]
[fv-az702-107:145232] [23] /opt/hostedtoolcache/Python/3.12.2/x64/lib/libpython3.12.so.1.0(PyImport_ImportModuleLevelObject+0x385)[0x7fe163894a85]
[fv-az702-107:145232] [24] /opt/hostedtoolcache/Python/3.12.2/x64/lib/libpython3.12.so.1.0(+0x279c93)[0x7fe163879c93]
[fv-az702-107:145232] [25] /opt/hostedtoolcache/Python/3.12.2/x64/lib/libpython3.12.so.1.0(+0x241ed3)[0x7fe163841ed3]
[fv-az702-107:145232] [26] /opt/hostedtoolcache/Python/3.12.2/x64/lib/libpython3.12.so.1.0(_PyEval_EvalFrameDefault+0x4f0e)[0x7fe163880e4e]
[fv-az702-107:145232] [27] /opt/hostedtoolcache/Python/3.12.2/x64/lib/libpython3.12.so.1.0(+0x219d35)[0x7fe163819d35]
[fv-az702-107:145232] [28] /opt/hostedtoolcache/Python/3.12.2/x64/lib/libpython3.12.so.1.0(PyObject_CallMethodObjArgs+0x103)[0x7fe16381ae43]
[fv-az702-107:145232] [29] /opt/hostedtoolcache/Python/3.12.2/x64/lib/libpython3.12.so.1.0(PyImport_ImportModuleLevelObject+0x6a9)[0x7fe163894da9]
[fv-az702-107:145232] *** End of error message ***
[fv-az702-107:00000] *** An error occurred in Socket closed
[fv-az702-107:00000] *** reported by process [3978166273,0]
[fv-az702-107:00000] *** on a NULL communicator
[fv-az702-107:00000] *** Unknown error
[fv-az702-107:00000] *** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
[fv-az702-107:00000] ***    and MPI will try to terminate your MPI job as well)
```

**Comment by dalcinl** [comment]
@jsquyres I this point I believe the whole intercomm and/or spawn thing is broken. In mpi4py, I disabled the high-stress spawn tests, but kept a few others that were not failing. Now you guys get new failures in this PR. A run of v5.0.x from today also failed unexpectedly [[link]](https://github.com/mpi4py/mpi4py-testing/actions/runs/8241699956/job/22548076522) at creating an intercomm (without spawn being involved).

**Comment by hppritcha** [comment]
> @hppritcha Is this the type of mpi4py spawn error that we should expect from this CI test? Or is this unexpected?
> 
> ```
> testErrcodes (test_spawn.TestSpawnMultipleWorld.testErrcodes) ... testErrcodes (test_spawn.TestSpawnMultipleWorld.testErrcodes) ... [fv-az702-107:145232] *** Process received signal ***
> [fv-az702-107:145232] Signal: Segmentation fault (11)
> [fv-az702-107:145232] Signal code: Address not mapped (1)
> ```

this is one of the failure modes.  note its generally not determistic.  for instance as @dalcinl points out its sometimes in ntercomm creation.  i've found that its best to try and run in an oversubscribed environment.  i resorted to a 4 cpu ubuntu vm .

**Comment by jsquyres** [comment]
Given that this is part of a known issue, I'm inclined to merge this PR anyway.  We should keep working the spawn issues separately.
